File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Actors/PlayActor.cs
using Zlink.Framework.Contracts.Actors;

namespace TicTacToe.Server.Play.Actors;

internal sealed class PlayActor(
    string actorId,
    IZLinkActorContext context)
    : IZLinkActor
{
    public string ActorId { get; } = actorId;

    public IZLinkActorContext Context { get; } = context;

    public string GameId { get; private set; } = string.Empty;

    public void JoinGame(string gameId)
    {
        if (string.IsNullOrWhiteSpace(gameId))
        {
            throw new ArgumentException("Game id must not be empty.", nameof(gameId));
        }

        GameId = gameId;
    }

    public string RequireJoinedGame()
    {
        if (!Context.IsJoined || string.IsNullOrEmpty(GameId))
        {
            throw new InvalidOperationException("Actor has not joined a game.");
        }

        return GameId;
    }

}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/adapters/zlink/actors/PlayActor.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.adapters.zlink.actors

import systems.zlink.framework.actors.ZLinkActor
import systems.zlink.framework.actors.ZLinkActorContext

class PlayActor(
    val actorId: String,
    private val context: ZLinkActorContext,
) : ZLinkActor {
    private var joinedRoomId: String? = null

    override fun actorId(): String = actorId

    override fun context(): ZLinkActorContext = context

    fun joinGame(roomId: String) {
        joinedRoomId = roomId
    }

    fun requireJoinedGame(): String =
        joinedRoomId ?: throw IllegalStateException("actor has not joined a game")
}
